File: Protótipos(back-end)/projetoFila(ts)/SistemaFila.ts
import Cliente from "./Cliente";
import ClienteInput from "./ClienteInput";
import Fila from "./Fila";
import Funcionario from "./Funcionario";
import FilaDisplay from ".FilaDisplay";
import * as readline from "readline";

class SistemaFila {
    private filasEspera: Map<number, Fila>;
    private senhasUtilizadas: Set<number>;
    private funcionarios: Funcionario[];
    private clienteInput: ClienteInput;
    private filaDisplay: FilaDisplay;

    constructor() {
        this.filasEspera = new Map<number, Fila>();
        this.senhasUtilizadas = new Set<number>();
        this.funcionarios = [];
        this.clienteInput = new ClienteInput();
        this.filaDisplay = new FilaDisplay();
    }

    gerarSenha(): number {
        let senha: number;
        do {
            senha = Math.floor(Math.random() * 10000);
        } while (this.senhasUtilizadas.has(senha));
        this.senhasUtilizadas.add(senha);
        return senha;
    }

    encerrarPrograma(senhaCorreta: string): boolean {
        const tentativasMaximas = 3;
        let tentativasRestantes = tentativasMaximas;
        const readline = require("readline").createInterface({
            input: process.stdin,
            output: process.stdout
        });

        const recursiveQuestion = () => {
            readline.question(`Digite a senha para encerrar o programa (Tentativas restantes: ${tentativasRestantes}): `, (senhaDigitada: string) => {
                if (senhaDigitada === senhaCorreta) {
                    console.log("Programa encerrado.");
                    readline.close();
                    return true;
                } else {
                    tentativasRestantes--;
                    console.log(`Senha incorreta. Tentativas restantes: ${tentativasRestantes}`);
                    if (tentativasRestantes > 0) {
                        recursiveQuestion();
                    } else {
                        console.log("Número de tentativas esgotado. Encerrando o programa.");
                        readline.close();
                        return false;
                    }
                }
            });
        };
        recursiveQuestion();
        return false;
    }

    adicionarCliente(nome: string, lugares: number): void {
        const fila = this.filasEspera.get(lugares) || new Fila(lugares);
        const senha = this.gerarSenha();
        try {
            const cliente = new Cliente(nome, senha);
            fila.adicionarCliente(cliente);
            this.filasEspera.set(lugares, fila);
        } catch (error) {
            console.log("Erro ao adicionar cliente.");
        }
    }

    exibirFilas(): void {
        this.filaDisplay.exibirFilas(this.filasEspera);
    }

    adicionarFuncionario(nome: string, cargo: string): void {
        const funcionario = new Funcionario(nome, cargo, FilaDisplay);
        this.funcionarios.push(funcionario);
    }

    iniciar(): void {
        const readline = require("readline").createInterface({
            input: process.stdin,
            output: process.stdout
        });

        const recursiveQuestion = () => {
            readline.question("Cliente: ", (nome: string) => {
                if (nome.toLowerCase() === "sair") {
                    if (this.encerrarPrograma("senha123")) {
                        readline.close();
                    } else {
                        recursiveQuestion();
                    }
                } else {
                    const clienteData = this.clienteInput.receberDadosCliente();
                    this.adicionarCliente(clienteData.nome, clienteData.lugares);
                    this.exibirFilas();
                    recursiveQuestion();
                }
            });
        };
        recursiveQuestion();
    }
}

export default SistemaFila;

File: Protótipos(back-end)/projetoFila(java)/SistemaFila.java
package projetoMascate;

import java.util.*;

public class SistemaFila {
    private Map<Integer, Fila> filasEspera;
    private Set<Integer> senhasUtilizadas;
    private List<Funcionario> funcionarios;
    private ClienteInput clienteInput;
    private FilaDisplay filaDisplay;

    public SistemaFila() {
        filasEspera = new HashMap<>();
        senhasUtilizadas = new HashSet<>();
        funcionarios = new ArrayList<>();
        clienteInput = new ClienteInput();
        filaDisplay = new FilaDisplay();
    }

    public Map<Integer, Fila> getFilasEspera() {
		return filasEspera;
	}

	public void setFilasEspera(Map<Integer, Fila> filasEspera) {
		this.filasEspera = filasEspera;
	}

	public Set<Integer> getSenhasUtilizadas() {
		return senhasUtilizadas;
	}

	public void setSenhasUtilizadas(Set<Integer> senhasUtilizadas) {
		this.senhasUtilizadas = senhasUtilizadas;
	}

	public List<Funcionario> getFuncionarios() {
		return funcionarios;
	}

	public void setFuncionarios(List<Funcionario> funcionarios) {
		this.funcionarios = funcionarios;
	}

	public ClienteInput getClienteInput() {
		return clienteInput;
	}

	public void setClienteInput(ClienteInput clienteInput) {
		this.clienteInput = clienteInput;
	}

	public FilaDisplay getFilaDisplay() {
		return filaDisplay;
	}

	public void setFilaDisplay(FilaDisplay filaDisplay) {
		this.filaDisplay = filaDisplay;
	}

	public int gerarSenha() {
        Random rand = new Random();
        int senha;
        do {
            senha = rand.nextInt(10000); // Gerando um número aleatório de até 4 dígitos
        } while (senhasUtilizadas.contains(senha));
        senhasUtilizadas.add(senha);
        return senha;
    }

    public boolean encerrarPrograma(String senhaCorreta) {
        Scanner scanner = new Scanner(System.in);
        int tentativasRestantes = 3;
        while (tentativasRestantes > 0) {
            System.out.print("Digite a senha para encerrar o programa (Tentativas restantes: " + tentativasRestantes + "): ");
            String senhaDigitada = scanner.nextLine();
            if (senhaDigitada.equals(senhaCorreta)) {
                System.out.println("Programa encerrado.");
                return true;
            } else {
                tentativasRestantes--;
                System.out.println("Senha incorreta. Tentativas restantes: " + tentativasRestantes);
            }
        }
        return false;
    }

    public void adicionarCliente(String nome, int lugares) {
        Fila fila = filasEspera.getOrDefault(lugares, new Fila(lugares));
        int senha = gerarSenha();
        try {
            Cliente cliente = new Cliente(nome, senha);
            fila.adicionarCliente(cliente);
            filasEspera.put(lugares, fila);
        } catch (IllegalArgumentException e) {
            System.out.println(e.getMessage());
        }
    }

    public void exibirFilas() {
        filaDisplay.exibirFilas(filasEspera);
    }

    public void adicionarFuncionario(String nome, String cargo) {
        Funcionario funcionario = new Funcionario(nome, cargo, filaDisplay);
        funcionarios.add(funcionario);
    }


    public final static void clearConsole()
    {
        try
        {
            final String os = System.getProperty("os.name");

            if (os.contains("Windows"))
            {
                Runtime.getRuntime().exec("cls");
            }
            else
            {
                Runtime.getRuntime().exec("clear");
            }
        }
        catch (final Exception e)
        {
        	System.out.println("Erro ao tentar limpar o console: " + e.getMessage());
        }
    }
    
    public void iniciar() {
        while (true) {
            ClienteInput.ClienteData clienteData = clienteInput.receberDadosCliente();
            if (clienteData.getNome().equalsIgnoreCase("sair")) {
                if (encerrarPrograma("senha123")) {
                    break;
                } else {
                    continue;
                }
            }
            adicionarCliente(clienteData.getNome(), clienteData.getLugares());
            exibirFilas();
        }
    }
}
